test/jasmine/tests/drawing_test.js

```diff
@@ -309,3 +315,83 @@ describe('Drawing', function() {
         });
     });
 });
+
+describe('gradients', function() {
+    var gd;
+
+    beforeEach(function() {
+        gd = createGraphDiv();
+    });
+
+    afterEach(destroyGraphDiv);
+
+    function checkGradientIds(uid, specs) {
+        var sortedExpected = specs.map(function(spec) {
+            return (spec[0] + uid + '-' + spec[1] + '-' + spec[2]).replace(/[^\w\-]+/g, '_');
+        }).sort();
+
+        var gids = [];
+        var gradients = d3.select(gd).selectAll('radialGradient,linearGradient');
+        gradients.each(function() { gids.push(this.id); });
+        gids.sort();
+
+        expect(gids.length).toBe(sortedExpected.length);
+
+        for(var i = 0; i < Math.min(gids.length, sortedExpected.length); i++) {
+            expect(gids[i]).toBe(sortedExpected[i]);
+        }
+    }
+
+    it('clears unused gradients after a replot', function(done) {
+        Plotly.plot(gd, [{
+            y: [0, 1, 2],
+            mode: 'markers',
+            marker: {
+                color: '#123',
+                gradient: {
+                    type: 'radial',
+                    color: ['#fff', '#eee', '#ddd']
+                }
+            }
+        }])
+        .then(function() {
+            checkGradientIds(gd._fullLayout._uid, [
+                ['radial', '#123', '#fff'],
+                ['radial', '#123', '#eee'],
+                ['radial', '#123', '#ddd']
+            ]);
+
+            return Plotly.restyle(gd, {'marker.color': '#456'});
+        })
+        .then(function() {
+            // simple scalar restyle doesn't trigger a full replot, so
+            // doesn't clear the old gradients
+            checkGradientIds(gd._fullLayout._uid, [
+                ['radial', '#123', '#fff'],
+                ['radial', '#123', '#eee'],
+                ['radial', '#123', '#ddd'],
+                ['radial', '#456', '#fff'],
+                ['radial', '#456', '#eee'],
+                ['radial', '#456', '#ddd']
+            ]);
```

Comment:
Not happy with this... there are a bunch of ways for it to blow up in actual use, like if someone makes 1000 different gradients via different marker colors, and animates them through 1000 intermediate states, suddenly you have a million gradients lying around.

I think I need to change it to associating the gradient with the point rather than with the color combo - unless someone has another idea how to keep this in check.